Root/OverlapTools.cxx

```diff
@@ -373,6 +385,287 @@ void OverlapTools::j_t_overlap(TauVector& taus, JetVector& jets, double dR)
+void OverlapTools_4Lep::photon_lep_overlap(PhotonVector& photons, ElectronVector& /*pre*/electrons, MuonVector& /*pre*/muons, double dR)
+{	
+  if( photons.size()==0) return;
+
+  // discard photons overlapping with muons
+  for(int iPho=photons.size()-1; iPho>=0; iPho--){
+    const Photon* pho = photons.at(iPho);
+    for(int iMu=muons.size()-1; iMu>=0; iMu--){
+      const Muon* mu = muons.at(iMu);
+
+      if(pho->DeltaRy(*mu) < dR){
+	photons.erase(photons.begin()+iPho);
+	break; 
+      } // if(dR< )
+
+    } // for(iMu)
+  } // for(iPho)
+
+  // discard photons overlapping with electrons
+  for(int iPho=photons.size()-1; iPho>=0; iPho--){
+    const Photon* pho = photons.at(iPho);
+    for(int iEl=electrons.size()-1; iEl>=0; iEl--) {
+      const Electron* el = electrons.at(iEl);
+
+      if(pho->DeltaRy(*el) < dR){
+	photons.erase(photons.begin()+iPho);
+	break; 
+      } // if(dR< )
+
+    } // for(iEl)
+  } // for(iPho)
+
+}
+
+//----------------------------------------------------------
+void OverlapTools_4Lep::j_e_overlap(ElectronVector& electrons, JetVector& jets, double dR)
+{
+  if(electrons.size()==0 || jets.size()==0) return;
+
+  for(int iEl=electrons.size()-1; iEl>=0; iEl--) {
+    const Electron* e = electrons.at(iEl);
+    for(int iJ=jets.size()-1; iJ>=0; iJ--){
+      const Jet* j = jets.at(iJ);
+
+      bool is_bjet = m_jetSelector->isB(j);
+      // only compare non-b-tagged jets
+      if( is_bjet ) continue;
+
+      if(e->DeltaRy(*j) < dR) {
+	if(verbose()) print_rm_msg("j_e_overlap: ", j, e);
+	jets.erase(jets.begin()+iJ);
+      }
+
+    } // iJ
+  } // iEl
+
+}
+
+//----------------------------------------------------------
+void OverlapTools_4Lep::e_j_overlap(ElectronVector& electrons, JetVector& jets, double dR)
+{
+  if(electrons.size()==0 || jets.size()==0) return;
+
+  for(int iEl=electrons.size()-1; iEl>=0; iEl--){
+    const Electron* e = electrons.at(iEl);
+    for(size_t iJ=0; iJ<jets.size(); iJ++){
+      const Jet* j = jets.at(iJ);
+      
+      bool pass_jvt = m_jetSelector->passJvt(j);
+      //don't compare with pileup jets
+      if( !m_useOldOverlap && !pass_jvt ) continue;
+      
+      if(e->DeltaRy(*j) < dR) {
+	if(verbose()) print_rm_msg("e_j_overlap: ", e, j);
+	electrons.erase(electrons.begin()+iEl);
+	break;
+      }
+      
+    } // for(iJ)
+  } // for(iEl)
+
+}
+
+//----------------------------------------------------------
+void OverlapTools_4Lep::j_m_overlap(MuonVector& muons, JetVector& jets, double dR)
+{
+  if(muons.size()==0 || jets.size()==0) return;
+
+  bool useGhostAssoc=true;
+  if(m_useOldOverlap) useGhostAssoc=false;
+
+  for(int iMu=muons.size()-1; iMu>=0; iMu--) {
+    const Muon* mu = muons.at(iMu);
+    for(int iJ=jets.size()-1; iJ>=0; iJ--){
+      const Jet* j = jets.at(iJ);
+
+      bool is_bjet = m_jetSelector->isB(j);
+      // only compare non-b-tagged jets
+      if( is_bjet ) continue;
+
+      bool discard_jet = false;
+      if(useGhostAssoc){
```

Comment:
Hi @tinapotter Comparing to AssociationUtils [1] it looks like the comparison between the pT ratios and nTracks is done whether or not the muon ghost association is being done. Here you only check these if you are not using the ghost association, no? Is this your analysis' intent?

[1] https://svnweb.cern.ch/trac/atlasoff/browser/PhysicsAnalysis/AnalysisCommon/AssociationUtils/tags/AssociationUtils-01-01-17/Root/MuJetOverlapTool.cxx#L160